sdk/endpoint/client.go

```diff
@@ -27,50 +27,47 @@ import (
 	"github.com/networkservicemesh/networkservicemesh/sdk/common"
 )
 
+// ClientEndpoint - opens a Client connection to another Network Service
 type ClientEndpoint struct {
-	BaseCompositeEndpoint
 	nsmClient     *client.NsmClient
 	mechanismType string
 	ioConnMap     map[string]*connection.Connection
 }
 
 // Request implements the request handler
+// Consumes from ctx context.Context:
+//	   Next
 func (cce *ClientEndpoint) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
-
-	if cce.GetNext() == nil {
-		logrus.Fatal("The connection composite requires that there is Next set.")
-	}
-
-	incomingConnection, err := cce.GetNext().Request(ctx, request)
-	if err != nil {
-		logrus.Errorf("Next request failed: %v", err)
-		return nil, err
-	}
-
-	var outgoingConnection *connection.Connection
 	name := request.GetConnection().GetId()
-	outgoingConnection, err = cce.nsmClient.Connect(name, cce.mechanismType, "Describe "+name)
+	outgoingConnection, err := cce.nsmClient.Connect(name, cce.mechanismType, "Describe "+name)
 	if err != nil {
 		logrus.Errorf("Error when creating the connection %v", err)
 		return nil, err
 	}
+	ctx = WithClientConnection(ctx, outgoingConnection)
+	incomingConnection := request.GetConnection()
+	if Next(ctx) != nil {
+		incomingConnection, err = Next(ctx).Request(ctx, request)
+		if err != nil {
+			return nil, err
+		}
+	}
 
-	// TODO: check this. Hack??
-	outgoingConnection.GetMechanism().GetParameters()[connection.Workspace] = ""
-
-	cce.ioConnMap[incomingConnection.GetId()] = outgoingConnection
+	cce.ioConnMap[request.GetConnection().GetId()] = outgoingConnection
```

Comment:
Should we save outgoingConnection into the map just after we created it to be able to close it even if next request will fail?